<h1 align="center">STRINGS</h1>
<h2 align="left"><ins>Lesson Guide</ins></h2>

- [**SPECIAL CHARACTERS**](#special)
- [**AUGMENTED ASSIGNMENT**](#augmented)
- [**STRING INDEXING & SLICING**](#index)
- [**BASIC BUILT-IN STRING METHODS**](#methods)
- [**is / in CHECK METHODS**](#isin)
- [**SPLIT, JOIN, REPLACE & STRIP**](#split)
- [**USER INPUT**](#userinput)
- [**STRING FORMATTING**](#format)

Strings are used in Python to record text information. Strings in Python are actually a *sequence*, which basically means Python keeps track of every element in the string as a sequence. For example, Python understands the string "hello' to be a sequence of letters in a specific order. This means we will be able to use indexing to grab particular letters (like the first letter, or the last letter).

- It's important to note that strings have an important property known as **immutability**. This means that once a string is created, the elements within it can not be changed or replaced.

In [1]:
print(dir(str))
# Returns an alphabetized list of names comprising the attributes
# of the given object, and of attributes reachable from it.

['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


In [2]:
help(str.capitalize)

Help on method_descriptor:

capitalize(self, /)
    Return a capitalized version of the string.
    
    More specifically, make the first character have upper case and the rest lower
    case.



<a id='special'></a>
## SPECIAL CHARACTERS
The [backslash](https://docs.python.org/3/reference/lexical_analysis.html#literals) (`\`) character is used to escape characters that otherwise have a special meaning, such as newline, backslash itself, or the quote character.

In [3]:
print("'Hello World'")
print('"hello world"')
print("\"Hello World\"")

print('I can\'t go \"there\"')
print("""I can't go "there". """)

# print("I can't go "there" ")  # results in an error

'Hello World'
"hello world"
"Hello World"
I can't go "there"
I can't go "there". 


In [4]:
print("""The pet shop owner said "No, no, 'e's uh,...he's resting".""")

print('The pet shop owner said "No, no, \'e\'s uh,...he\'s resting".')
print("The pet shop owner said \"No, no, 'e's uh,...he's resting\".")

The pet shop owner said "No, no, 'e's uh,...he's resting".
The pet shop owner said "No, no, 'e's uh,...he's resting".
The pet shop owner said "No, no, 'e's uh,...he's resting".


In [5]:
print("""This string has 
been split
over
several lines""")
print()
print("""This string has not \
been split \
over \
several lines""")

This string has 
been split
over
several lines

This string has not been split over several lines


In [6]:
# print("C:\Users\timbuchalka\notes.txt")  
# Running the above line will result in an error because of \U (reference to unicode)

# An alternative approach is to use \\ or the raw-string literal
print("C:\\Users\\timbuchalka\\notes.txt")
print(r"C:\Users\timbuchalka\notes.txt")    

C:\Users\timbuchalka\notes.txt
C:\Users\timbuchalka\notes.txt


- `\033[1m` turns the string into bold.
- `\033[0m` removes the bold. 

In [7]:
print('\033[1mHello')
print(f'\033[1mHello\033[0mworld')
print(f'\033[1mHello\033[0m world')

Hello
Helloworld
Hello world


The <code>center()</code> method allows you to place your string 'centered' between a provided string with a certain length. Personally, I've never actually used this in code as it seems pretty esoteric...

In [8]:
s = 'hello world'
s.center(20,'-')

'----hello world-----'

The <code>expandtabs()</code> method will expand tab notations <code>\t</code> into spaces:

In [9]:
'hello\tworld'.expandtabs()

'hello   world'

In [10]:
'hello\tworld'

'hello\tworld'

In [11]:
print('hello\tworld')

hello	world


In [12]:
print('hello\nworld')

hello
world


<a id='augmented'></a>
## AUGMENTED ASSIGNMENT
Just like with numbers, we can use **augmented assignment** with strings.

In [13]:
x=5
x+=1
print(x)

greeting = "Good "
greeting += "Morning!"
print(greeting)       # similar to concatenation

greeting *= 5
print(greeting)       # useful way to print out strings multiple times

6
Good Morning!
Good Morning!Good Morning!Good Morning!Good Morning!Good Morning!


In [14]:
# You can add strings together to concatenate them (join them together).
name = "Jose"
greeting = "Hello, " + name

print(greeting)

# You cannot add strings and numbers, as this will result in an error:
# age = 34
# print("You are " + age)    

# But you can if you turn the number into a string first:
age = 34
age_as_string = str(age)
print("You are " + age_as_string)

Hello, Jose
You are 34


<a id='index'></a>
## STRING INDEXING & SLICING

Remember that positive indexing starts at 0, while negative indexing begins at -1.

In [15]:
string = "Hello World!"

# returns the letter at index 6
print(string[6])           

# returns the second last letter
print(string[-2])  

# another way to get the negative index from a positive one.
print(string[6-12])     

W
d
W


In [16]:
string = "Hello World!"

# returns the letters from index range 0 to 6 (inclusive)
print(string[:7])          

# returns the letters from index range 7 to the end
print(string[7:])        

# returns the letters in the index range 3 to 6 (inclusive)
print(string[3:7])

# returns World! be cautious though using pos and neg together
print(string[-6:12])    

# returns World
print(string[-6:-1])

Hello W
orld!
lo W
World!
World


In [17]:
string = ''
print(string)       # this will work
print(string[:1])   # this will work
# print(string[0])   # this will produce an error

In [18]:
# General form is [start:stop:step]

print(string[3:9:2])    # start at 3, ends at 6, every second letter

print(string[::-1])     # returns the string backwards

print(string[12:0:-1])  # returns string backwards without first letter
print(string[12::-1])   # this returns the string backwards as well

In [19]:
# Challenge
letters = 'abcdefghijklmnopqrstuvwxyz'

# 1) Create a slice that produces the characters qpo
print(letters[-10:-13:-1])
print(letters[16:13:-1])

# 2) Slice the string to produce edcba
print(letters[-22::-1])
print(letters[4::-1])

# 3) Slice the string to produce the last 8 characters, in reverse order
print(letters[:-9:-1])
print(letters[26:17:-1])

qpo
qpo
edcba
edcba
zyxwvuts
zyxwvuts


<a id='methods'></a>
## BASIC BUILT-IN STRING METHODS
Documentation: https://docs.python.org/3/library/stdtypes.html#string-methods<br>
Useful resource: https://www.w3schools.com/python/python_ref_string.asp

Objects in Python usually have built-in methods. These methods are functions inside the object that can perform actions or commands on the object itself. String objects have a variety of methods we can use to save time and add functionality.    

We call methods with a period and then the method name:<br>
`object.method(parameters)`<br>
where parameters are extra arguments we can pass into the method. 

In [20]:
string = "Hello World!"

# returns the index number for the first occurence of 'o'
print(string.index("o"))  

# if we needed the second 'o' we could do
print(string.index('W')+1) 

# returns the number of times the letter 'l' occurs
print(string.count('l')) 

# returns the number of characters (including spaces, etc...)
print(len(string))        

4
7
3
12


In [21]:
# returns the starting index position of the first occurence
string.find('lo') 

3

We can use the multiplication symbol to create repetition!

In [22]:
letter = 'z'
letter*10

'zzzzzzzzzz'

In [23]:
# produces an error since strings are immutable and cannot 
# be changed once created.

s = 'Hello'
s[0] = 'Y'     # we cannot change the 'H' to a 'Y'
print(s)

TypeError: 'str' object does not support item assignment

In [24]:
print(s.replace('H', 'Y'))
print(s)

Yello
Hello


In [25]:
astring = "Hello World!"
print(astring.upper())     # Convert all the string characters to CAPITAL letters
print(astring.lower())     # Convert the string to lowercase letters 
print(astring.casefold())  # Converts all characters to lower case

print(astring.startswith("Hello"))       # checks if the string starts with 'hello' and returns True or False. s[0]
print(astring.endswith("asdfasdfasdf"))  # same as above but checks the end of the string. s[-1]

HELLO WORLD!
hello world!
hello world!
True
False


In [26]:
astring.capitalize()  # sets the first letter of the string to a capital

'Hello world!'

In [27]:
astring.title()  # capitalizes the first letter of each word in a string

'Hello World!'

In [28]:
astring.swapcase()  # inverts the casing for each element in the string

'hELLO wORLD!'

In [29]:
def palindrome_check(word):
    # check against full sequence in reverse order
    if word == word[::-1]:  
        return True
    #else:
    #    return False
    return False

print(palindrome_check("kayak"))  # True
print(palindrome_check("lemon"))  # False

True
False


**Remember, strings are immutable. None of the above methods change the string in place, they only return modified copies of the original string.**

<a id='isin'></a>
## is / in CHECK METHODS
These various methods below check if the string is some case. Let's explore them:

In [30]:
s = 'hello'

# returns True if all characters in s are alphanumeric:
print(s.isalnum())

# returns True if all characters in s are alphabetic:
print(s.isalpha())

# returns True if all cased characters in s are lowercase (extends to uppercase as well):
print(s.islower())

# returns True if all characters in s are whitespace (i.e an empty string):
print(s.isspace())

# returns True if s is a title cased string:
print(s.istitle())

True
True
True
False
False


In [31]:
# returns True if the string is a numeric string, False otherwise.
print(s.isnumeric())

# A string is numeric if all characters in the string are numeric and there is at
# least one character in the string.

# returns True if the string is a digit string, False otherwise.
print(s.isdigit())

False
False


In [32]:
'h' in s

True

In [33]:
'h' not in s

False

In [34]:
print('x' in [1,2,3])        #False

print('x' in ['x','y','z'])  #True

print('x' in ['xyz'])        #False

print('x' in 'xyz')          #True

False
True
False
True


<a id='split'></a>
## SPLIT, JOIN, REPLACE & STRIP
Strings have some built-in methods that can resemble regular expression operations.
We can use <code>split()</code> to split the string at a certain element and return a list of the results.
We can use <code>partition()</code> to return a tuple that includes the first occurrence of the separator sandwiched between the first half and the end half.

In [35]:
bstring = "   Hello World   "
bstring.strip()          #removes all blank spaces on either ends of the string

'Hello World'

In [36]:
cstring = "$$$Hello World$$$"
print(cstring.strip('$')) #only removes $ at the beginning or end of a string    

print(cstring.lstrip('$'))  # only removes from the left hand side

print(cstring.rstrip('$'))  # only removes from the right hand side

Hello World
Hello World$$$
$$$Hello World


In [37]:
print(bstring.replace(" ", '') ) #locates every ' ' in the string a replaces it with ''

print(bstring.replace('H', 'Y'))

HelloWorld
   Yello World   


In [38]:
astring = "hello world!"

print(astring.split())   # Splits a string by blank space (this is the default)

# Splits at all instances by a specific element (doesn't include the element that was split on)
print(astring.split('e'))

['hello', 'world!']
['h', 'llo world!']


In [39]:
#splits at the first instance
astring.partition('l')

('he', 'l', 'lo world!')

In [40]:
'_'.join(astring)   #use _ to join all elements in astring

'h_e_l_l_o_ _w_o_r_l_d_!'

In [41]:
test = 'hello'
test.split()

['hello']

In [42]:
'.'.join(test).split('.')

['h', 'e', 'l', 'l', 'o']

In [43]:
new_string = []
for word in astring.split():
    new_string.append(word.capitalize())
    
' '.join(new_string)

'Hello World!'

In [44]:
' '.join(word.capitalize() for word in s.split())

'Hello'

In [45]:
number = "9,223;372:036 854,775;807"
seperators = number[1::4]     #starts at index 1 (inclusive) and then counts 4 steps until the end
print(seperators)

#this first determine whether each char is a number or a seperator, then splits into a list of strings.
values = "".join(char if char not in seperators else " " for char in number).split()

print(values)
print([int(val) for val in values])  #converts the values from a string into an integer

,;: ,;
['9', '223', '372', '036', '854', '775', '807']
[9, 223, 372, 36, 854, 775, 807]


In [46]:
# Challenge

# Consider the string: my_string = 'a0:12:90:00:80:43'.
# Write a code that will return a list consisting of the elements of this string which are delimited by a colon,
# using a string-specific method.

my_string = 'a0:12:90:00:80:43'
print(my_string.split(':'))
print(''.join(my_string.split(':')))

['a0', '12', '90', '00', '80', '43']
a01290008043


<a id='userinput'></a>
## USER INPUT

We use input() to present the user with a prompt. 
They can then type, and when they press 'Enter' everything they typed gets returned.
We can assign that to a variable if we want!
Just remember that `input()` by itself will always return the users response as a string.

In [47]:
name = "Bob"
your_name = input("Enter your name: ")

print(f"Hello {your_name}. My name is {name}.")

Enter your name: barry
Hello barry. My name is Bob.


In [48]:
## Calculating months

age = input("Enter your age: ")  # Enter 3
print(f"You have lived for {age * 12} months.")  

Enter your age: 35
You have lived for 353535353535353535353535 months.


In [49]:
"""
It doesn't work quite as we'd expect.

That's because we're multiplying the string '3' by 12, which concatenates the string twelve times.
Instead we want to multiply the number. We'll have to convert our string to a number first.
"""

age = input("Enter your age: ")  
age_num = int(age)
print(f"You have lived for {age_num * 12} months.") 

Enter your age: 35
You have lived for 420 months.


In [50]:
"""
We could do it in one line, a bit more simply.
After all, `age` doesn't need to be a string at any point.
"""

age = int(input("Enter your age: "))  
print(f"You have lived for {age * 12} months.")  

Enter your age: 35
You have lived for 420 months.


In [51]:
"""
A better option would be to create a `months` variable, as that is cleaner to read.
"""

age = int(input("Enter your age: "))  
months = age * 12
print(f"You have lived for {months} months.")  

Enter your age: 35
You have lived for 420 months.


<a id='format'></a>
## STRING FORMATTING

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. 

There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method. 
* The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

### Formatting with placeholders
You can use <code>%s</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".

In [52]:
# This prints out "John is 23 years old."
name = "John"
age = 23
print("%s is %d years old." % (name, age))

John is 23 years old.


In [53]:
# This prints out "Hello, John!"
names = ['john', 'michael', 'benji']
for name in names:
    print("Hello, %s!" % name)

print('The names on the list are %s' % names)
print('The names on the list are', names[:])

print('The second name on the list is', names[1])
print('The second name on the list is %s' % names[1])

Hello, john!
Hello, michael!
Hello, benji!
The names on the list are ['john', 'michael', 'benji']
The names on the list are ['john', 'michael', 'benji']
The second name on the list is michael
The second name on the list is michael


In [54]:
# Here are some basic argument specifiers you should know:

# %s - String (or any object with a string representation, like numbers)

# %d - Integers

# %f - Floating point numbers

# %.<number of digits>f - Floating point numbers with a fixed amount of digits to the right of the dot.

# %x/%X - Integers in hex representation (lowercase/uppercase)

In [55]:
data = ("John", "Doe", 53.44)
format_string = "Hello %s %s. Your current balance is $%s."

print(format_string % data)

Hello John Doe. Your current balance is $53.44.


In [56]:
favorite_cars = ['landcruiser', 'tacoma', 'obs', 'jeep']
message = f"My first car was a {favorite_cars[2].upper()}."
print(message)

favorite_cars = ['landcruiser', 'tacoma', 'obs', 'jeep']
message = "My first car was a {favorite_cars[2].upper()}."
print(message)

My first car was a OBS.
My first car was a {favorite_cars[2].upper()}.


It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. We will learn more about these functions later on in the course, but you should note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [57]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


As another example, `\t` inserts a tab into a string.

In [58]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first, without rounding. Note the difference below:

In [59]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


### Multiple Formatting
Nothing prohibits using more than one conversion tool in the same print statement:

In [60]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


### Padding and Precision of Floating Point Numbers
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

In [61]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [62]:
print('Floating point numbers: %1.0f' %(13.144))

Floating point numbers: 13


In [63]:
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [64]:
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [65]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


In [66]:
data = ("John", "Doe", 53.44)

print('Hello %s %s. Your current balance is $%.2f.' % data)   #the .2 refers to the number of decimal places

Hello John Doe. Your current balance is $53.44.


### Formatting with the `.format()` method
A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [67]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


In [68]:
name = 'Michael'
final_greeting = "How are you, {}?".format(name)
print(final_greeting)

# The {} gets replaced by whatever is inside the brackets of the .format()

How are you, Michael?


In [69]:
# You can also give names to variables inside a formattable string:

friend_name = "Rolf"
goodbye = "Goodbye, {name}!"
goodbye_rolf = goodbye.format(name=friend_name)
print(goodbye_rolf)


# Another example of using `.format()` on a variable:

greeting = "How are you, {}?"
print(greeting.format(name))

# Usually you will be using f-strings, just because they are shorter and more readable.
# However sometimes you may need to re-use a format string, and that is when using .format() is useful.

Goodbye, Rolf!
How are you, Michael?


In [70]:
description = '{} is {} years old.'
print(description.format('Bob', 30))

Bob is 30 years old.


In [71]:
description = '{} is {age} years old.'
print(description.format('Bob', age=30))

Bob is 30 years old.


### The .format() method has several advantages over the %s placeholder method:

#### 1. Inserted objects can be called by index position:

In [72]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


In [73]:
print("Jan: {2}, Feb: {0}, Mar: {2}, Apr: {1}, May: {2}, Jun: {1}, Jul: {2}, Aug: {2}, Sep: {1}, Oct: {2}, \
Nov: {1}, Dec: {2}".format(28, 30, 31))

Jan: 31, Feb: 28, Mar: 31, Apr: 30, May: 31, Jun: 30, Jul: 31, Aug: 31, Sep: 30, Oct: 31, Nov: 30, Dec: 31


In [74]:
print("""Jan: {2}, Feb: {0}, Mar: {2}, 
Apr: {1}, May: {2}, Jun: {1}, Jul: {2}, 
Aug: {2}, Sep: {1}, Oct: {2},
Nov: {1}, Dec: {2}""".format(28, 30, 31))

Jan: 31, Feb: 28, Mar: 31, 
Apr: 30, May: 31, Jun: 30, Jul: 31, 
Aug: 31, Sep: 30, Oct: 31,
Nov: 30, Dec: 31


#### 2. Inserted objects can be assigned keywords:

In [75]:
for i in range(1,13):
    print("No. {0} squared is {1} and cubed is {2}".format(i, i**2, i**3))

No. 1 squared is 1 and cubed is 1
No. 2 squared is 4 and cubed is 8
No. 3 squared is 9 and cubed is 27
No. 4 squared is 16 and cubed is 64
No. 5 squared is 25 and cubed is 125
No. 6 squared is 36 and cubed is 216
No. 7 squared is 49 and cubed is 343
No. 8 squared is 64 and cubed is 512
No. 9 squared is 81 and cubed is 729
No. 10 squared is 100 and cubed is 1000
No. 11 squared is 121 and cubed is 1331
No. 12 squared is 144 and cubed is 1728


In [76]:
for i in range(1,13):
    print("No. {0:2} squared is {1:3} and cubed is {2:4}".format(i, i**2, i**3))

No.  1 squared is   1 and cubed is    1
No.  2 squared is   4 and cubed is    8
No.  3 squared is   9 and cubed is   27
No.  4 squared is  16 and cubed is   64
No.  5 squared is  25 and cubed is  125
No.  6 squared is  36 and cubed is  216
No.  7 squared is  49 and cubed is  343
No.  8 squared is  64 and cubed is  512
No.  9 squared is  81 and cubed is  729
No. 10 squared is 100 and cubed is 1000
No. 11 squared is 121 and cubed is 1331
No. 12 squared is 144 and cubed is 1728


In [77]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


#### 3. Inserted objects can be reused, avoiding duplication:

In [78]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


### Alignment, padding and precision with `.format()`
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [79]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:8}'.format('Apples', 3.))
print('{0:8} | {1:8}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |      3.0
Oranges  |       10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [80]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character

In [81]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [82]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

In [83]:
print("Pi is approximately {0}".format(22/7))
print("Pi is approximately {0:12}".format(22/7))
print("Pi is approximately {0:52.50f}".format(22/7))
print("Pi is approximately {0:62.50f}".format(22/7))
print("Pi is approximately {0:72.50f}".format(22/7))
print("Pi is approximately {0:<72.50f}".format(22/7))
print("Pi is approximately {0:<72.54f}".format(22/7))

Pi is approximately 3.142857142857143
Pi is approximately 3.142857142857143
Pi is approximately 3.14285714285714279370154144999105483293533325195312
Pi is approximately           3.14285714285714279370154144999105483293533325195312
Pi is approximately                     3.14285714285714279370154144999105483293533325195312
Pi is approximately 3.14285714285714279370154144999105483293533325195312                    
Pi is approximately 3.142857142857142793701541449991054832935333251953125000                


### Formatted String Literals (f-strings)

Introduced in Python 3.6, f-strings offer several benefits over the older `.format()` string method described above. For one, you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`.

In [84]:
# f-strings don't work in Python 3.5 or earlier.
# In f-strings, {name} gets replaced by the value of the variable name.

another_greeting = f"How are you, {name}?"
print(another_greeting)

How are you, Michael?


In [85]:
name = 'Fred'

print(f"He said his name is {name}.")

He said his name is Fred.


Pass `!r` to get the string representation:

In [86]:
print(f"He said his name is {name!r}")

He said his name is 'Fred'


In [87]:
name = 'rolf smith'
street = '123 no name road'
postcode = 'py10 1cp'

address = f'''Name: {name!r}
Street: {street}
Postcode: {postcode}
Country: United Kingdom'''

print(address)

Name: 'rolf smith'
Street: 123 no name road
Postcode: py10 1cp
Country: United Kingdom


#### Float formatting follows `"result: {value:{width}.{precision}}"`

Where with the `.format()` method you might see `{value:10.4f}`, with f-strings this can become `{value:{10}.{6}}`


In [88]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")
print(f"My 10 character, four decimal number is: {num:.6f}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is: 23.456780


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal. This fits more closely with scientific notation and statistical analysis. Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [89]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use `.format()` method syntax inside an f-string:

In [90]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


In [91]:
number = 234.23493450295
print(f'my number to 3 decimals is: {number:.3f}.')

my number to 3 decimals is: 234.235.


In [92]:
for i in range(1,13):
    print(f"No. {i:2} squared is {i**2:3} and cubed is {i**3:4}")

No.  1 squared is   1 and cubed is    1
No.  2 squared is   4 and cubed is    8
No.  3 squared is   9 and cubed is   27
No.  4 squared is  16 and cubed is   64
No.  5 squared is  25 and cubed is  125
No.  6 squared is  36 and cubed is  216
No.  7 squared is  49 and cubed is  343
No.  8 squared is  64 and cubed is  512
No.  9 squared is  81 and cubed is  729
No. 10 squared is 100 and cubed is 1000
No. 11 squared is 121 and cubed is 1331
No. 12 squared is 144 and cubed is 1728


In [93]:
for i in range(1,13):
    print(f"No. {i:2} squared is {i**2:<3} and cubed is {i**3:^4}")

No.  1 squared is 1   and cubed is  1  
No.  2 squared is 4   and cubed is  8  
No.  3 squared is 9   and cubed is  27 
No.  4 squared is 16  and cubed is  64 
No.  5 squared is 25  and cubed is 125 
No.  6 squared is 36  and cubed is 216 
No.  7 squared is 49  and cubed is 343 
No.  8 squared is 64  and cubed is 512 
No.  9 squared is 81  and cubed is 729 
No. 10 squared is 100 and cubed is 1000
No. 11 squared is 121 and cubed is 1331
No. 12 squared is 144 and cubed is 1728
